proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy

```diff
@@ -0,0 +1,727 @@
+  and cur_thread[wp]: "\<lambda>s. P (cur_thread s)"
+  (wp: mapM_x_wp_inv thread_set.arch_state select_wp crunch_wps
+   simp: crunch_simps possible_switch_to_def reschedule_required_def)
+
+lemma do_unbind_notification_arch_tcb_at[wp]:
+  "do_unbind_notification ntfnptr ntfn tcbptr \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding set_bound_notification_def set_simple_ko_def
+  apply (wpsimp wp: set_object_wp get_object_wp get_simple_ko_wp thread_get_wp)
+  by (fastforce simp: pred_tcb_at_def obj_at_def get_tcb_def)
+
+lemma unbind_notification_arch_tcb_at[wp]:
+  "unbind_notification tcb \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding unbind_notification_def
+  by wpsimp
+
+lemma bind_notification_arch_tcb_at[wp]:
+  "bind_notification tcbptr ntfnptr \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding bind_notification_def set_bound_notification_def set_simple_ko_def
+  apply (wpsimp wp: set_object_wp get_object_wp get_simple_ko_wp)
+  by (fastforce simp: pred_tcb_at_def obj_at_def get_tcb_def)
+
+lemma unbind_maybe_notification_arch_tcb_at[wp]:
+  "unbind_maybe_notification ntfnptr \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding unbind_maybe_notification_def
+  by wpsimp
+
+crunches blocked_cancel_ipc, cap_delete_one, cancel_signal
+  for arch_tcb_at[wp]: "arch_tcb_at P t"
+  (wp: crunch_wps simp: crunch_simps)
+
+lemma reply_cancel_ipc_arch_tcb_at[wp]:
+  "reply_cancel_ipc ntfnptr \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding reply_cancel_ipc_def thread_set_def
+  apply (wpsimp wp: set_object_wp select_wp)
+  by (clarsimp simp: pred_tcb_at_def obj_at_def get_tcb_def)
+
+crunches cancel_ipc, send_ipc, receive_ipc
+  for arch_tcb_at[wp]: "arch_tcb_at P t"
+  (wp: crunch_wps simp: crunch_simps)
+
+lemma send_fault_ipc_arch_tcb_at[wp]:
+  "send_fault_ipc tptr fault \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding send_fault_ipc_def thread_set_def Let_def
+  apply (rule validE_valid)
+  apply (rule hoare_seq_ext_skipE, solves wpsimp)+
+  apply (wpsimp wp: set_object_wp)
+  apply (clarsimp simp: pred_tcb_at_def obj_at_def get_tcb_def)
+  done
+
+crunches handle_fault, handle_interrupt, handle_vm_fault, handle_hypervisor_fault, send_signal
+  for arch_tcb_at[wp]: "arch_tcb_at P t"
+  (wp: valid_cur_vcpu_lift mapM_x_wp_inv crunch_wps)
+
+crunches reschedule_required, set_scheduler_action, tcb_sched_action
+  for arch_tcb_at[wp]: "arch_tcb_at P t"
+  (simp: reschedule_required_def set_scheduler_action_def tcb_sched_action_def set_tcb_queue_def
+         get_tcb_queue_def)
+
+lemma do_reply_transfer_arch_tcb_at[wp]:
+  "do_reply_transfer sender receiver slot grant \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding do_reply_transfer_def thread_set_def
+  apply (rule hoare_seq_ext_skip, solves wpsimp, clarsimp?)+
+  apply (rename_tac fault_opt)
+  apply (case_tac fault_opt; clarsimp)
+   apply wpsimp
+  apply (rule hoare_seq_ext_skip, solves wpsimp)+
+  apply (wpsimp wp: set_object_wp hoare_drop_imps | intro conjI impI)+
```

Comment:
The `intro conjI impI` is not a good pattern in combination with `wpsimp`, it's usually a sign that something has gone wrong. In this case it's splitting an `if` into a conjunction instead of progressing with `wp`.

Try 
`  apply (wpsimp wp: set_object_wp hoare_drop_imps split_del: split_if)`
instead (note no `+`, which `wpsimp` should never need for a single goal).

This specific thing (too eager `if_split`) is a problem with the `clarsimp` inside `wpsimp` being too eager. We could try to remove `if_split` in there by default, but we'd get lots of annoying warnings in contexts where `if_split` is already removed. Maybe we should put this one on the list of things to spend some time thinking about.

Reply:
Dropping the `if_split` is one way, and `hoare_drop_imps` is useful when you don't need to the condition of the `if`. If you *do* need the condition of the `if`, consider adding the simp rule `if_apply_def2` which converts `if` to implications, then you can decide whether you want to drop them or not. I don't know if this applies here, but it's a very useful tool to know about.

Reply:
The `if` is in the code this time (somewhere deep in a term), so this specific rule wouldn't help here, but in general it's a good idea to check, yes.